lib/srv/regular/sshserver.go

```diff
@@ -1096,40 +1109,254 @@ func (s *Server) getServerResource() (types.Resource, error) {
+	// to the desired endpoint.
+	dialer := func(addr string) (net.Conn, error) {
+		local, remote, err := uds.NewSocketpair(uds.SocketTypeStream)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		defer remote.Close()
+
+		remoteFD, err := remote.File()
+		if err != nil {
+			local.Close()
+			return nil, trace.Wrap(err)
+		}
+		defer remoteFD.Close()
+
+		_, _, err = proc.Conn.WriteWithFDs([]byte(addr), []*os.File{remoteFD})
+		if err != nil {
+			local.Close()
+			return nil, trace.Wrap(err)
+		}
+
+		return local, nil
+	}
+
+	// try to register with the parent context.
+	if other, ok := scx.Parent().TrySetDirectTCPIPForwardDialer(dialer); !ok {
+		// another forwarder was concurrently created. this isn't actually a problem, multiple forwarders
+		// being registered is harmless, but it does result in slightly higher resource utilization, so its
+		// preferable to use the existing forwarder and close ours in the background.
+		go proc.Close()
+		return other, nil
+	}
+
+	// successfully registered this dialer, add closer to context.
+	scx.Parent().AddCloser(proc)
+
+	return dialer, nil
+}
+
+// listenTCPIP creates a new listener in the forwarding process.
+func (s *Server) listenTCPIP(scx *srv.ServerContext, addr string) (net.Listener, error) {
+	proc, err := s.getTCPIPForwardProcess(scx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Create a socket to receive new connections on.
+	localConn, remoteConn, err := uds.NewSocketpair(uds.SocketTypeDatagram)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer localConn.Close()
+	defer remoteConn.Close()
+	remoteFD, err := remoteConn.File()
+	if err != nil {
+		localConn.Close()
+		return nil, trace.Wrap(err)
+	}
+	defer remoteFD.Close()
+	_, _, err = proc.Conn.WriteWithFDs([]byte(addr), []*os.File{remoteFD})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// The forwarding process may have chosen its own port, so we need to get the
+	// new listen address.
+	var fn int
+	fbuf := make([]*os.File, 1)
+	readDone := make(chan struct{})
+	// Read addr in another goroutine so we can cancel it if the forwarding process
+	// stops.
+	go func() {
+		_, fn, err = localConn.ReadWithFDs(nil, fbuf)
+		close(readDone)
+	}()
+
+	select {
+	case <-proc.Done:
+		localConn.Close()
+		return nil, trace.Errorf("forwarding process is closed")
+	case <-readDone:
+	}
+
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if fn == 0 {
+		return nil, trace.BadParameter("forwarding process did not return a listener")
+	}
+	if err := validateListenerSocket(scx, localConn.UnixConn, fbuf[0]); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	listener, err := net.FileListener(fbuf[0])
+	return listener, trace.Wrap(err)
+}
+
+func controlSyscallConn(conn syscall.RawConn, f func(fd uintptr) error) error {
+	var err error
+	err2 := conn.Control(func(fd uintptr) {
+		err = f(fd)
+	})
+	return trace.NewAggregate(err2, err)
```

Comment:
In all sensible cases (i.e. a `RawConn` for a real file) `Control` will only return an error before calling the function if the file isn't valid:
```suggestion
	var err error
	if cErr := conn.Control(func(fd uintptr) {
		err = f(fd)
	}); cErr != nil {
		return trace.Wrap(cErr)
	}
	return trace.Wrap(err)
```